vcd/resource_vcd_vapp_static_routing.go

```diff
@@ -0,0 +1,214 @@
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+	vcdClient.lockParentVappWithName(d, vapp.VApp.Name)
+	defer vcdClient.unLockParentVappWithName(d, vapp.VApp.Name)
+
+	networkId := d.Get("network_id").(string)
+	staticRouting, err := expandVappNetworkStaticRouting(d)
+	if err != nil {
+		return fmt.Errorf("error expanding static routes: %s", err)
+	}
+	vappNetwork, err := vapp.UpdateNetworkStaticRouting(networkId, staticRouting, d.Get("enabled").(bool))
+	if err != nil {
+		log.Printf("[INFO] Error setting static routing: %s", err)
+		return fmt.Errorf("error setting static routing: %s", err)
+	}
+
+	d.SetId(vappNetwork.ID)
+
+	return resourceVappNetworkStaticRoutingRead(d, meta)
+}
+
+func resourceVAppNetworkStaticRoutingDelete(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vcdClient.lockParentVappWithName(d, vapp.VApp.Name)
+	defer vcdClient.unLockParentVappWithName(d, vapp.VApp.Name)
+
+	err = vapp.RemoveAllNetworkStaticRoutes(d.Get("network_id").(string))
+	if err != nil {
+		log.Printf("[INFO] Error deleting static routes: %s", err)
+		return fmt.Errorf("error deleting static routes: %s", err)
+	}
+
+	return nil
+}
+
+func resourceVappNetworkStaticRoutingRead(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vappNetwork, err := vapp.GetVappNetworkById(d.Get("network_id").(string), false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp network. %s", err)
+	}
+
+	var rules []map[string]interface{}
+	if vappNetwork.Configuration.Features == nil || vappNetwork.Configuration.Features.StaticRoutingService == nil {
+		log.Print("[INFO] no Static routes found.")
+		_ = d.Set("rule", nil)
+	}
+
+	for _, rule := range vappNetwork.Configuration.Features.StaticRoutingService.StaticRoute {
+		singleRule := make(map[string]interface{})
+
+		singleRule["name"] = rule.Name
+		singleRule["network_cidr"] = rule.Network
+		singleRule["next_hop_ip"] = rule.NextHopIP
+		rules = append(rules, singleRule)
+	}
+	_ = d.Set("enabled", vappNetwork.Configuration.Features.StaticRoutingService.IsEnabled)
```

Comment:
I think I have found a bug. Enabled is **not** set during the first apply. Reproduce:

1.) Rename a static routing resource, you'll get this as expected:

```
Terraform will perform the following actions:

  # vcd_vapp_static_routing.vapp1_static_routing will be destroyed
  - resource "vcd_vapp_static_routing" "vapp1_static_routing" {
      - enabled    = true -> null
      - id         = "urn:vcloud:network:bd0c0774-4ed3-4e42-ba5d-94f4bc13f8da" -> null
      - network_id = "urn:vcloud:network:bd0c0774-4ed3-4e42-ba5d-94f4bc13f8da" -> null
      - vapp_id    = "urn:vcloud:vapp:a62383a1-0140-409a-a60e-2dad7ba4e49f" -> null

      - rule {
          - name         = "rule1" -> null
          - network_cidr = "10.10.0.0/24" -> null
          - next_hop_ip  = "192.168.0.2" -> null
        }
      - rule {
          - name         = "rule2" -> null
          - network_cidr = "10.10.1.0/24" -> null
          - next_hop_ip  = "192.168.0.4" -> null
        }
    }

  # vcd_vapp_static_routing.vapp_demo_static_routing will be created
  + resource "vcd_vapp_static_routing" "vapp_demo_static_routing" {
      + enabled    = true
      + id         = (known after apply)
      + network_id = "urn:vcloud:network:bd0c0774-4ed3-4e42-ba5d-94f4bc13f8da"
      + vapp_id    = "urn:vcloud:vapp:a62383a1-0140-409a-a60e-2dad7ba4e49f"

      + rule {
          + name         = "rule1"
          + network_cidr = "10.10.0.0/24"
          + next_hop_ip  = "192.168.0.2"
        }
      + rule {
          + name         = "rule2"
          + network_cidr = "10.10.1.0/24"
          + next_hop_ip  = "192.168.0.4"
        }
    }

Plan: 1 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes
```

Run `terraform apply` for the second time, you'll get:

```
# vcd_vapp_static_routing.vapp_demo_static_routing will be updated in-place
  ~ resource "vcd_vapp_static_routing" "vapp_demo_static_routing" {
      ~ enabled    = false -> true
        id         = "urn:vcloud:network:bd0c0774-4ed3-4e42-ba5d-94f4bc13f8da"
        network_id = "urn:vcloud:network:bd0c0774-4ed3-4e42-ba5d-94f4bc13f8da"
        vapp_id    = "urn:vcloud:vapp:a62383a1-0140-409a-a60e-2dad7ba4e49f"

      + rule {
          + name         = "rule1"
          + network_cidr = "10.10.0.0/24"
          + next_hop_ip  = "192.168.0.2"
        }
      + rule {
          + name         = "rule2"
          + network_cidr = "10.10.1.0/24"
          + next_hop_ip  = "192.168.0.4"
        }
    }
```

Please check whether FW and NAT has same issue as well.

CC: @dataclouder @Didainius

Reply:
vcd_vapp_static_routing.rul1: Destroying... [id=urn:vcloud:network:ca0b8d64-3568-4e89-b38e-d083f14fc2d9]
vcd_vapp_static_routing.rul11: Creating...
vcd_vapp_firewall_rules.rul1: Modifying... [id=urn:vcloud:network:ca0b8d64-3568-4e89-b38e-d083f14fc2d9]
vcd_vapp_static_routing.rul11: Creation complete after 6s [id=urn:vcloud:network:ca0b8d64-3568-4e89-b38e-d083f14fc2d9]
vcd_vapp_static_routing.rul1: Still destroying... [id=urn:vcloud:network:ca0b8d64-3568-4e89-b38e-d083f14fc2d9, 10s elapsed]
vcd_vapp_firewall_rules.rul1: Still modifying... [id=urn:vcloud:network:ca0b8d64-3568-4e89-b38e-d083f14fc2d9, 10s elapsed]
vcd_vapp_firewall_rules.rul1: Modifications complete after 12s [id=urn:vcloud:network:ca0b8d64-3568-4e89-b38e-d083f14fc2d9]
vcd_vapp_static_routing.rul1: Destruction complete after 16s

It's not a bug, it happens that descrution is last thing and result creation was overriden

Reply:
OK, so Terraform tries to destroy and create in parallel. Then because this is a flat representation in VCD API (unlike other resources which each has its own ID/place), and because destroy, through a race-condition, happened last, it wiped out the configuration. In other words, renaming in Terraform needs to be treated with special care.